ifan02 cookbook

quazzie opened the pull request:
## Description:
Sonoff ifan02 cookbook.
It is not pretty but the info is correct.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#<esphome PR number goes here>
**Pull request in [esphome-core](https://github.com/esphome/esphome-core) with C++ framework changes (if applicable):** esphome/esphome-core#<esphome-core PR number goes here>

## Checklist:

  - [ ] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.

quazzie committed:
ifan02 cookbook

quazzie committed:
removed spaces

quazzie committed:
removed spaces

quazzie committed:
indent codeblock, newline at end

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+Ifan02
```
Longer title please (include sonoff and fix capitalization)

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+
+.. seo::
+    :description: Instructions for using Sonoff ifan02 in ESPHome.
+    :keywords: Fan, Sonoff, ifan02
```
This needs an `:image:` too.

quazzie replied:
Hmm where to find a good free non copyrighted image ? Mine are installed so i can't take a picture.

quazzie replied:
i just used the fan.svg

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+    :keywords: Fan, Sonoff, ifan02
+
+
+Configuration for `Sonoff ifan02 <https://www.itead.cc/sonoff-ifan02-wifi-smart-ceiling-fan-with-light.html>`__.
```
Longer description please. Explain what the iFan02 is, what it does and what functionality the user will get in the end.

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+
+Configuration for `Sonoff ifan02 <https://www.itead.cc/sonoff-ifan02-wifi-smart-ceiling-fan-with-light.html>`__.
+
+First you need a :doc:`components/output/custom.html` to control the ifan02.
```
```suggestion
First you need a :doc:`components/output/custom` to control the iFan02.
```

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+    #include "esphome.h"
+    using namespace esphome;
+
+    class IFan02Output : public Component, public output::FloatOutput {
```
```suggestion
    class IFan02Output : public output::FloatOutput {
```

this component has no setup/loop function, so it doesn't need any component functionality.

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+              digitalWrite(4, LOW);
+              digitalWrite(15, HIGH);
+            }
+        }
```
rewrite this to be
```cpp
if (state < 0.3) {
  // OFF
} else if (state < 0.6) {
  // low speed
} else if (state < 0.9) {
  // medium speed
} else {
  // high speed
}
```

There are some values that this component does not handle like 0.31 (the if/else if fixes that kinda)

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+
+      - platform: gpio
+        pin: GPIO15
+        id: fan_relay3
```
What all this gpio switches here when the custom output already handles those? This could potentially cause issues in the future if: for example the output platform writes the state more often and other stuff.

The custom output should be the only controller of these GPIOs. Please remove this logic and use `fan.turn_on` / `fan.turn_off` in the button triggers.

quazzie replied:
This is because when you use the remote you get virtual button presses (vbutton_*) and you should toggle those relays. Depending on what speed you are on and what speed you want to go to diffrent vbutton fire. Pressing 1 on the remote will not fire the same vbuttons all the time. 
For example from pressing 1 on remote when fan is off the remote fires vbutton1, but pressing 1 when fan is on medium will fire vbutton2. 
And the vbuttons need to toggle corresponding relays.

OttoWinter commented on cookbook/ifan02.rst:
```diff
@@ -0,0 +1,181 @@
+      - platform: speed
+        output: fanoutput
+        id: ifan02
+        name: ifan02_fan
```
A see also section is missing here. It should point to some of the components used here (fan, custom output, etc)

OttoWinter reviewed (commented):
- If you don't add this file to index.rst, it won't show up on the front page. It needs to be added there in the cookbook section.

quazzie committed:
fixes

quazzie committed:
add seo :image:

quazzie committed:
spaces...

CallMeAreks commented:
Hi. Is this example correct? I have the ifan02 already mounted so I wanted to confirm if it's working as intended before flashing via OTA.

workingmanrob commented:
Hey @CallMeAreks 

I've been using Quazzie's cookbook for a couple months now without issue.

I was running Tasmota firmware on my sonoff fan02 so I downloaded the esphome firmware after I built it then used the tasmota web UI to flash and everything is working as expected.

CallMeAreks commented:
@workingmanrob Thank you for the confirmation. I tried it as well like a week ago, (leap of faith because I have 4 already installed ifan02), but work flawlessly. Forgot to leave a comment, though.

pstechsolutions commented:
Not sure what made my case different.  I had an issue during compile where I would get an error about only being allowed to register sub-components.  The solution: In the  ifan02.h file the line 

class IFan02Output : public output::FloatOutput {

needed to be

class IFan02Output : public Component, public output::FloatOutput {

@OttoWinter and @quazzie Thank you both for your efforts and making them available for those of us, like myself, who are just learning!

CallMeAreks commented:
@pstechsolutions Yeah, now that you mention this I noticed the code in this commit history is slightly different from the one I copied: https://gist.github.com/quazzie/09ee3ef2c419ecbcf979a7410062481b

In this gist it reflects the change you made. Leaving the link here for future reference.

workingmanrob commented:
Updated to 1.13.2 am now getting a valication error:

Unknown value 'MEDIUM', did you mean 'MEDIuM'?

Pretty sure I mean MEDIUM and trying to use MEDIuM instead gave the same error.

Any ideas?

workingmanrob commented:
Looks like 1.13.3 fixed it.. thanks Otto!

mjoshd committed:
Complete the example code & fix compilation errors

Tried compiling this code today using ESPHome 1.13.6.. it would not compile. 

* The ifan02.h file was the main culprit and I was able to fix the issue.

* The .yaml example was missing a few declarations at the beginning so I added them.
* I fixed an error related to using ```light``` for the light id since that is a reserved name.
* I added an ```on_boot``` declaration to turn off the light as early as possible after power is restored. (if mains power is restored in the middle of the night and the light turns on while you are sleeping it can be quite annoying, this turns it off during the boot process).

quazzie committed:
Merge pull request #1 from mjoshd/patch-2

Complete the example code & fix compilation errors

mjoshd committed:
removed blank spaces on line 66

Those blank spaces get me every time ;-)

quazzie committed:
Merge pull request #2 from mjoshd/patch-3

removed blank spaces on line 66

mjoshd committed:
fix doc links (hopefully)

I think my last PR was made against the wrong fork/branch/etc... sorry, I'm still learning github.
I currently have the time and really want to see this make it into the cookbook. I think if you add me as a collaborator I would be able to push the changes and help get the errors sorted more quickly; if not, no worries, I'll take no offense and do my best to continue helping as we are doing now.

mjoshd committed:
Merge pull request #4 from mjoshd/patch-1

fix doc links (in the right repo)

mjoshd committed:
fix fan link

replace
```
:doc:`/components/fan`
```
with
```
:doc:`/components/fan/index`
```

mjoshd commented:
All is good now!

@quazzie Thanks for letting me assist!

@OttoWinter I have successfully compiled and installed this on my iFan02 with ESPHome 1.13.6 and it works 100% perfectly. Please merge this into the docs so others may benefit from it as well. 

Have a great day everyone! 😃

quazzie commented:
It's only stale because @OttoWinter has not responded.

quazzie commented:
Still not stale @OttoWinter

quazzie commented:
I guess it must be stale as nothing is happening...

mjoshd commented:
It appears that a link has been added in the DIY Examples
<https://esphome.io/guides/diy.html#sample-configurations> section but it's
pointing to your original gist and not to the code we worked on in the PR.

We should try to get that link re-pointed to the code from the PR at some
point.




On Mon, Mar 16, 2020 at 2:34 AM Tommy Jonsson <notifications@github.com>
wrote:

> I guess it must be stale as nothing is happening...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/esphome/esphome-docs/pull/218#issuecomment-599367841>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIQ77RK4YKKRBZLA2GPWLGDRHXB6XANCNFSM4HDTTG2Q>
> .
>

glmnet committed:
Merge branch 'current' into ifan02
